### PR TITLE
Implement select widget

### LINKED ISSUE #140: Select widget
As a person editing the site metadata, I'd like to have single and multiple choice select widgets so I can see what the options are and pick from a list.

#### Acceptance Criteria:

- [ ] Implement widget using react-select or any other component that supports selecting multiple items
- [ ] Validation on frontend
- [ ] Schema validation should support the new widget. See schema here: https://docs.google.com/document/d/1qmUEoiYupPJlftdadgYe0jN6WoMnO0IdmQ99WuQQLuw/edit

#### Out of Scope

- Loading options from a remote source ("Advanced" in the google doc)

### PR DESCRIPTION
**Pull request by noisecapella**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #140 

#### What's this PR do?
Implements the select widget and some frontend validation for the `min` and `max` field settings.

#### How should this be manually tested?

 - Edit `site-config-override.yml` and add these two lines to the metadata fields list (should be the very last lines of the file at the moment):

        - { "label": "Tags", "default": ["Design"], "max": 3, "min": 1, "multiple": true, "name": "tags", "options": ["Design", "UX", "Dev"], "widget": "select" }
        - { label: "Align Content", name: "align", widget: "select", options: ["left", "center", "right"]}
        
 - Go to `/admin/websites/websitestarter/1/change/` or a similar URL for your WebsiteStarter. Update the config textbox to include the new line you added above. If you are using the default WebsiteStarter called "OCW Course Hugo Starter" you can just copy the paste the contents from `site-config-override.yml` minus the `collections:` line at the top.
 - Go to a page for a site which uses that starter. Click on "Site Metadata" on the sidebar then click "Add Site Metadata". You should see a field called "Tags", and it should have a value "Design" set as a default option.
 - Update the items in the Tags select box. You should be able to add and remove items. Click Save and they should be saved correctly.
 - You should also be able to update the Align Content field, though you should not be able to add or remove multiple items.
 - Go back to the Site Metadata page and click Edit. You should see the fields you updated with the new values.
 - Edit your metadata entry again and remove all selected items from the "Tags" field. Click Save. You should see an error message saying that there must be at least 1 item for that field.

#### Screenshots (if appropriate)
Screenshot from 2021-03-24 16-57-57

### CODE REVIEW

File: websites/serializers.py [hunk 1]
```diff
@@ -157,7 +157,68 @@ class Meta:
...
+                    raise ValidationError(f"Expected string for {field_name}")
+            elif widget == "select":
+                if field.get("multiple"):
+                    if not isinstance(value, list):
+                        raise ValidationError(f"Expected list for {field_name}")
+                else:
+                    if not isinstance(value, (str, type(None))):
+                        raise ValidationError(
+                            f"Expected string or null for {field_name}"
+                        )
```
Comment (gsidebo) [inline websites/serializers.py:216]: Most/all of these validations seem to be for request data that should not be possible coming from the front end. I can think of a scenario where the data we get does not match the config fields – a user loads a content form with some site config, that config is updated on the backend, then the user submits the form. That's a larger issue that we need to address comprehensively, though. Adding this validation is jumping the gun, IMO. Lmk if I'm missing something

File: websites/factories.py [hunk 2]
```diff
@@ -31,7 +21,14 @@ class WebsiteStarterFactory(DjangoModelFactory):
...
     source = FuzzyChoice(STARTER_SOURCES)
     commit = factory.Faker("md5")
-    config = EXAMPLE_SITE_CONFIG
+    config = factory.LazyAttribute(
+        lambda _: yaml.load(
+            (
+                Path(settings.BASE_DIR) / "localdev/starters/site-config-override.yml"
+            ).read_text(),
+            Loader=yaml.Loader,
+        )["course"]
```
Comment (gsidebo) [inline websites/factories.py:30]: Use `COURSE_STARTER_SLUG`

**Review by gsidebo — CHANGES_REQUESTED** [review]

**Comment by aliceriot** [comment]
I'm seeing a validation issue where it says 'Tags is a required field' even though there is a value there:

Screenshot from 2021-03-30 13-12-56
Screenshot from 2021-03-30 13-12-23

I used the `override_site_configs` management command.

**Comment by aliceriot** [comment]
word, just tried it and it looks good. I think we don't want tags to be a required field anyhow.

File: static/js/components/widgets/SelectField.test.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+      const wrapper = shallow(
+        <SelectField
+          value={null}
+          name={name}
+          onChange={sandbox.stub()}
+          options={options}
+          multiple={true}
+        />
+      )
```
Comment (aliceriot) [inline static/js/components/widgets/SelectField.test.tsx:122]: could be helpful to write a simple `render` function to DRY up the test cases a little bit

File: static/js/components/widgets/SelectField.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  const { value, onChange, name, options } = props
+  const multiple = Boolean(props.multiple)
+
+  const changeHandler = (newValue: any) => {
+    const eventValue = multiple ?
+      newValue.map((option: Option) => option.value) :
+      newValue.value
+    // @ts-ignore
+    onChange({ target: { value: eventValue, name } })
+  }
```
Comment (aliceriot) [inline static/js/components/widgets/SelectField.tsx:29]: probably a good idea to wrap this function with `useCallback`

File: static/js/components/widgets/SelectField.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  } else {
+    selected = isNil(value) ? null : { label: value, value: value }
+  }
+
+  return (
+    <Select
+      value={selected}
+      isMulti={Boolean(multiple)}
+      onChange={changeHandler}
+      // @ts-ignore
```
Comment (aliceriot) [inline static/js/components/widgets/SelectField.tsx:47]: I don't see any errors if I delete this `@ts-ignore`

File: static/js/lib/site_content.ts [hunk 6]
```diff
@@ -38,13 +41,13 @@ const isMainContentField = (field: ConfigField) =>
  * Translates page content form values into a payload that our REST API understands.
  */
 export const contentFormValuesToPayload = (
-  values: { [key: string]: string | File },
+  values: Record<string, any>,
```
Comment (aliceriot) [inline static/js/lib/site_content.ts:44]: is there any way we can avoid `any` here?

Comment (aliceriot) [reply static/js/lib/site_content.ts:44]: I changed it to `Record<string, string | File>` and didn't get any type errors except for one in a test file

**Review by aliceriot — CHANGES_REQUESTED** [review]
a few comments / questions

**Review by aliceriot — APPROVED** [review]
lgtm! :+1:

**Pull request merged**